src/vsr/replica.zig

```diff
@@ -8761,6 +9092,88 @@ pub fn ReplicaType(
             // responsibility to check.
             self.send_message_to_replica(parameters.replica, reply);
         }
+
+        fn send_commit(self: *Self) void {
+            assert(self.status == .normal);
+            assert(self.primary());
+            assert(self.commit_min == self.commit_max);
+
+            if (self.primary_abdicating) {
+                assert(self.primary_abdicate_timeout.ticking);
+                assert(self.pipeline.queue.prepare_queue.count > 0);
+                assert(self.primary_pipeline_pending() != null);
+
+                log.mark.debug("{}: send_commit: primary abdicating (view={})", .{
+                    self.replica,
+                    self.view,
+                });
+                return;
+            }
+
+            const latest_committed_entry = checksum: {
+                if (self.commit_max == self.superblock.working.vsr_state.checkpoint.commit_min) {
+                    break :checksum self.superblock.working.vsr_state.checkpoint.commit_min_checksum;
+                } else {
+                    break :checksum self.journal.header_with_op(self.commit_max).?.checksum;
+                }
+            };
+
+            self.send_header_to_other_replicas_and_standbys(@bitCast(Header.Commit{
+                .command = .commit,
+                .cluster = self.cluster,
+                .replica = self.replica,
+                .view = self.view,
+                .commit = self.commit_max,
+                .commit_checksum = latest_committed_entry,
+                .timestamp_monotonic = self.clock.monotonic(),
+                .checkpoint_op = self.superblock.working.vsr_state.checkpoint.commit_min,
+                .checkpoint_id = self.superblock.working.checkpoint_id(),
+            }));
+        }
+
+        fn send_request_upgrade_to_self(self: *Self) void {
+            assert(self.status == .normal);
+            assert(self.primary());
+            assert(self.upgrade_release.? > self.release);
+            maybe(self.pipeline.queue.message_by_operation(.upgrade) != null);
+
+            const upgrade = vsr.UpgradeRequest{ .release = self.upgrade_release.? };
+            self.send_request_to_self(.upgrade, std.mem.asBytes(&upgrade));
+
+            // If our clock is not synchronized, then the request-to-self might have been dropped.
+            maybe(self.pipeline.queue.message_by_operation(.upgrade) != null);
+        }
+
+        fn send_request_to_self(self: *Self, operation: vsr.Operation, body: []const u8) void {
```

Comment:
I wrote this thinking that the [pulse messages](https://github.com/tigerbeetle/tigerbeetle/pull/1657) might be able to use this same function, but it looks like pulses are injected in a very different way.